test_workspace/TreeList.bosatsu

```diff
@@ -0,0 +1,152 @@
+package TreeList
+
+export [ TreeList, empty, cons, decons, head, get, fold, eq_TreeList ]
+
+# implementation of O(1) cons/uncons O(log N) list-like data structure
+
+enum Tree:
+  Single(a: a)
+  Branch(size: Int, head: a, left: Tree[a], right: Tree[a])
+
+def foldTree(t: Tree[a], init: b, fn: b -> a -> b) -> b:
+  # performance is better if we minimize the scope.
+  # this is a good candidate for optimization to fix
+  # this issue...
+  def loop(t: Tree[a], init: b) -> b:
+    recur t:
+      Single(a): fn(init, a)
+      Branch(_, h, l, r):
+        init = fn(init, h)
+        init = loop(l, init)
+        loop(r, init)
+  loop(t, init)
+
+struct TreeList(trees: List[Tree[a]])
+
+empty: forall a. TreeList[a] = TreeList([])
+
+operator + = add
+
+def cons(head: a, tail: TreeList[a]) -> TreeList[a]:
+  TreeList(trees) = tail
+  match trees:
+    []: TreeList([Single(head)])
+    [s1 @ Single(_)]:
+      TreeList([Single(head), s1])
+    [s1 @ Single(_), b1 @Branch(_, _, _, _), *t]:
+      TreeList([Single(head), b1, *t])
+    [s1 @ Single(_), s2 @ Single(_), *t]:
+      TreeList([Branch(3, head, s1, s2), *t])
+    [Branch(_, _, _, _), Single(_), *t]:
+      # illegal state, we can just return empty
+      empty
+    [branch @ Branch(_, _, _, _)]: TreeList([Single(head), branch])
+    [Branch(_, _, _, _), Single(_), *t]: empty
+    [b1@Branch(s1, _, _, _), b2@Branch(s2, _, _, _), *t]:
+      if (eq_Int(s1, s2)):
+        TreeList([Branch(s1 + s2 + 1, head, b1, b2), *t])
+      else:
+        TreeList([Single(head), *trees])
+
+def decons(tail: TreeList[a]) -> Option[(a, TreeList[a])]:
+  TreeList(trees) = tail
+  match trees:
+    []: None
+    [Single(h), *t]: Some((h, TreeList(t)))
+    [Branch(_, h, b1, b2), *t]: Some((h, TreeList([b1, b2, *t])))
+
+def head(tl: TreeList[a]) -> Option[a]:
+  match decons(tl):
+    Some((h, _)): Some(h)
+    None: None
+
+operator + = add
+operator - = sub
+
+#def int_loop(intValue: Int, state: a, fn: Int -> a -> (Int, a)) -> a
```

Comment:
what does `#def` mean?

Reply:
Oh... this is just a comment so the signature is handy

Reply:
Yep. I should remove.